README.md
# catClicker
Become the greatest crazy cat lady!
Maak zo veel mogelijk magische katten.
Maar pas op, als je er te veel maakt zullen de consequenties heel naar zijn....


Het project maakt gebruik van deze OOP principes.

• Classes
------------------------------------------------------------------------------------------------------------------------------------
Waar is het toegepast, en waarom is het op die plek toegepast:

"Classes inherit functionality and objects are built from these classes"

De Class witch wordt in de playscreen.ts aangeroepen en wordt daar een x en y coordinaat mee gegeven.
Hierdoor kun je meerdere witches aanroepen met andere x en y coordinaten in de playscreen.ts zonder this.div.style.transform te gebruiken. Omdat de witch class deze al doorgeeft aan de playscreen.ts.

------------------------------------------------------------------------------------------------------------------------------------
code voorbeelden: Witch.ts:

class Witch extends GameObject {
    constructor(el: any, x: number, y: number) {
        super(el, x, y);
        this.update()
    }
    public update() {
        this.div.style.transform = `translate(${this.x}px, ${this.y}px`
    }
}


Playscreen.ts:
------------------------------------------------------------------------------------------------------------------------------------

class PlayScreen {
  public game: Game;
  public witch: Witch

constructor(g: Game) {
  this.game = g;
  //create witch
  this.witch = new Witch("witch", 500, 337)
 

• Encapsulation
------------------------------------------------------------------------------------------------------------------------------------

Waar is het toegepast, en waarom is het op die plek toegepast:
Ik heb encapsulation toegepast in de playscreen.ts. Alle objecten die ik alleen in de playscreen.ts gebruik staan op private, zodat andere classes ze niet aan kunnen roepen. Wanneer ik onderdelen van classes door andere classes aan wil roepen, zoals de witch class onderdelen in de playscreen, dan gebruik ik public.

Encapsulation betekent zoveel als inkapseling. In OO is elk object afgeschermd. Objecten kunnen elkaar niet 'zien'. Ze kunnen wel allerlei dingen uitwisselen, alsof ze via een luikje met elkaar in verbinding staan.
code voorbeelden:

private - Only the current class will have access to the field or method.

protected - Only the current class and subclasses (and sometimes also same-package classes) of this class will have access to the field or method.

public - Any class can refer to the field or call the method.

class PlayScreen {
  public game: Game
  public witch: Witch
  public cat: Cat
  public count: Counter
  private tooMuchCats: number = 25
  public background: Background
  private ypos: number
  private xpos: number
  private screenW: number
  private screenH: number
  private randomNumW: number
  private randomNumH: number
  
  constructor(g: Game) {
    this.screenW = window.innerWidth - 320
    this.screenH = window.innerHeight - 296
    this.randomNumW = Math.random()
    this.randomNumH = Math.random()
    this.xpos = (this.screenW * this.randomNumW)
    this.ypos = (this.screenH * this.randomNumH)
    this.game = g

    //----------------------------->create background
   
   this.background = new Background("background", 0, 0)
   this.witch = new Witch("witch", 500, 337)
   this.witch.div.addEventListener("click", () => this.counter())
   this.count = new Counter("counter", this.screenW/2, 100)

   //create empty cat
  
   this.cat = new Cat("empty", 0,0)

   counter() {
    this.randomNumW = Math.random();
    this.randomNumH = Math.random()
    this.xpos = (this.screenW * this.randomNumW);
    this.ypos = (this.screenH * this.randomNumH);
    this.cat = new Cat("cat", this.xpos, this.ypos)
    this.cat.div.addEventListener("click", () => this.deleteCat())
    //----------------------------->create cat
    this.count.num++  // this.counter++; 
    console.log("1up")
    this.count.div.innerHTML = ("You have " + this.count.num + " cats");//add text 2 counter
    //add eventlistener delete cat

  }
  deleteCat() {
    this.count.num--;
    this.count.div.innerHTML = ("You have " + this.count.num + " cats");//add text 2 counter
   
  }

  clickedSoMuch() {  //if cat tag > then 45 = game Over;
    this.game.emptyScreen();
    this.game.showGameOver(new GameOver(this.game))
  }

  public update() {
   
   if (this.count.num > this.tooMuchCats) { this.clickedSoMuch() };
  }
}


• Composition 
------------------------------------------------------------------------------------------------------------------------------------
uitleg: waar is het toegepast, en waarom is het op die plek toegepast:

Ik heb een witch, counter en cat class die ik aanroep in de playscreen.ts. 

class PlayScreen {
public game: Game

  public witch: Witch
  public cat: Cat
  public count: Counter
  private tooMuchCats: number = 25
  public background: Background
  
  
------------------------------------------------------------------------------------------------------------------------------------
Hierdoor kan ik de variabelen samen in een functie gebruiken in de playscreen.ts


counter() {

   this.randomNumW = Math.random()
   this.randomNumH = Math.random()
   this.xpos = (this.screenW * this.randomNumW)
   this.ypos = (this.screenH * this.randomNumH)
   this.cat = new Cat("cat", this.xpos, this.ypos)
   this.cat.div.addEventListener("click", () => this.deleteCat())  
   this.count.num++   
   console.log("1up")
   this.count.div.innerHTML = ("You have " + this.count.num + " cats");//add text 2 counter

  }
  deleteCat() {
    this.count.num--;
    this.count.div.innerHTML = ("You have " + this.count.num + " cats");//add text 2 counter
   
  }

  clickedSoMuch() {  //if cat  > then 25 = game Over
    this.game.emptyScreen();
    this.game.showGameOver(new GameOver(this.game))
  }
  

• Inheritance 
------------------------------------------------------------------------------------------------------------------------------------
Waar is het toegepast, en waarom is het op die plek toegepast:
Ik heb een gameobject gemaakt die met inheritance  variabelen en functies doorgeeft aan de witch, de counter en de background.ts
Hierdoor kun je meerdere objecten divs maken met verschilende x en y coordinaten in de playscreen.ts zonder voor elke class een aparte div te creeren en x en y coordinaat te verklaren, omdat deze al doorgegeven wordt door de gameobject.ts.

code voorbeelden:

gameobject.ts
------------------------------------------------------------------------------------------------------------------------------------
class GameObject {
    public x: number
    public y: number
    public div: HTMLElement

   constructor(el: any, x: number, y: number) {
        this.x = x
        this.y = y
        this.div = document.createElement(el)
        document.body.appendChild(this.div)
        

witch.ts:
------------------------------------------------------------------------------------------------------------------------------------
class Witch extends GameObject {
    constructor(el: any, x: number, y: number) {
        super(el, x, y);
        this.update()
    }
    public update() {
        this.div.style.transform = `translate(${this.x}px, ${this.y}px`
    }
}


Playscreen.ts:
------------------------------------------------------------------------------------------------------------------------------------
class PlayScreen {
  public game: Game;
  public witch: Witch
 
  constructor(g: Game) {

  this.game = g
  this.witch = new Witch("witch", 500, 337)
 
------------------------------------------------------------------------------------------------------------------------------------

 • Een klassendiagram van de game.
https://github.com/madameJeanette/catClicker/blob/master/Class%20Diagram%20cat%20Clicker.png
 
 
 • Een link naar de peer review die in week 6 is gedaan. 
 https://github.com/madameJeanette/fluffyGameWork/issues/2
 
 • Een link naar online versie
 https://stud.hosted.hr.nl/0872675/websites/catClicker/public/
 


howler.js/CHANGELOG.md
## 2.0.12 (May 9, 2018)
- `FIXED` The previous Chrome deprecation fixes broke spatial positioning in Safari.

## 2.0.10 (May 5, 2018)
- `FIXED` Fixed another Chrome deprecation warning when using panning methods ([#923](https://github.com/goldfire/howler.js/issues/923)).
- `FIXED` Playback rate wasn't working correctly in Internet Explorer when defined in the `Howl` constructor ([#936](https://github.com/goldfire/howler.js/issues/936)).
- `FIXED` Looped audio would only play twice in Internet Explorer ([#921](https://github.com/goldfire/howler.js/issues/921)).

## 2.0.9 (February 10, 2018)
- `FIXED` More accurate HTML5 Audio `end` timer and fix for Firefox streams ending early ([#883](https://github.com/goldfire/howler.js/issues/883)).
- `FIXED` Prevent `play` events from duplicating in certain instances ([#899](https://github.com/goldfire/howler.js/issues/899)).
- `FIXED` Add second parameter to HTML5 Audio playback promise to fix Safari error ([#896](https://github.com/goldfire/howler.js/pull/896)).
- `FIXED` Refactored the internal queue system to fix various edge cases.

## 2.0.8 (January 19, 2018)
- `CHANGED` Fades now use elapsed time to be more accurate when intervals are inconsistent ([#885](https://github.com/goldfire/howler.js/issues/885)).
- `CHANGED` Improve timing of short fades ([#884](https://github.com/goldfire/howler.js/issues/884)).
- `FIXED` Fixed another Chrome deprecation when setting playback rate.
- `FIXED` Prevent `onplay` from firing when first setting `stereo` value ([#843](https://github.com/goldfire/howler.js/issues/843)).

## 2.0.7 (December 18, 2017)
- `FIXED` Accidental `const` was included in the previous version.

## 2.0.6 (December 15, 2017)
- `FIXED` Replaced deprecated `gain.value` and `gain.pan.value` with `setValueAtTime` ([#856](https://github.com/goldfire/howler.js/issues/856)).
- `FIXED` Audio sprites weren't ending correctly in Internet Explorer 11 ([#841](https://github.com/goldfire/howler.js/issues/841)).
- `FIXED` Correctly set group volume when fading ([#539](https://github.com/goldfire/howler.js/issues/539)).
- `FIXED` Cancel `fade` on sound when `mute` is called ([#666](https://github.com/goldfire/howler.js/issues/666)).
- `FIXED` Uncaught error when play() request was interrupted by a call to pause() ([#835](https://github.com/goldfire/howler.js/pull/835)).
- `FIXED` Incorrect reference to global `_scratchBuffer` ([#834](https://github.com/goldfire/howler.js/pull/834)).

## 2.0.5 (October 6, 2017)
- `ADDED` Add support for `withCredentials` to Web Audio XHR requests ([#610](https://github.com/goldfire/howler.js/pull/610)).
- `ADDED` Add `playerror` event for when mobile HTML5 audio is unable to play ([#774](https://github.com/goldfire/howler.js/issues/774)).
- `FIXED` Refactor fade method to eliminate bind memory allocations (no change to API).
- `FIXED` Prevent seeking after sound has been unloaded ([#797](https://github.com/goldfire/howler.js/pull/797)).
- `FIXED` Check for `paused` instead of `ended` on HTML5 end check to correctly handle data URI's ([#775](https://github.com/goldfire/howler.js/pull/775)).
- `FIXED` Fix unlocking of mobile audio on iOS when user swipes instead of taps ([#808](https://github.com/goldfire/howler.js/pull/808)).
- `FIXED` `pannerAttr` values can now be set via object as the documentation originally specified.
- `FIXED` Various corrections and improvements to the spatial audio documentation.

## 2.0.4 (June 9, 2017)
- `CHANGED` Removed the `resuming` state, which wasn't actually being used and was leading to a bug on Android ([#679](https://github.com/goldfire/howler.js/pull/679)).
- `CHANGED` Any playback initiated before the sound has loaded will now go into the queue to fix various race conditions ([#714](https://github.com/goldfire/howler.js/pull/714)).
- `FIXED` Correctly initialize an AudioContext with the global mute status ([#714](https://github.com/goldfire/howler.js/pull/714)).
- `FIXED` AudioContext unlocks on user interaction within a cross-domain iframe on Android Chrome ([#756](https://github.com/goldfire/howler.js/pull/756)).
- `FIXED` Stopping/pausing a group of sounds now behaves as expected in edge cases ([#734](https://github.com/goldfire/howler.js/pull/734)).
- `FIXED` Sound ID's now start at 1000 instead of 0 to avoid `rate` collisions ([#764](https://github.com/goldfire/howler.js/issues/764)).
- `FIXED` Prevent unknown mime errors on Internet Explorer when unloading a sound ([#720](https://github.com/goldfire/howler.js/pull/720)).
- `FIXED` Correctly clean up error event listeners ([#720](https://github.com/goldfire/howler.js/pull/720)).
- `FIXED` Audio clipping in Internet Explorer when network latency is present with HTML5 Audio ([#720](https://github.com/goldfire/howler.js/pull/720)).
- `FIXED` Allow passing just an event and ID to turn off listener ([#767](https://github.com/goldfire/howler.js/issues/767)).
- `FIXED` `npm` warning caused by invalid license definition ([#763](https://github.com/goldfire/howler.js/pull/763)).

## 2.0.3 (March 11, 2017)
- `CHANGED` Unloading a sound no longer fires the `end` event ([#675](https://github.com/goldfire/howler.js/pull/675)).
- `FIXED` Remove `setTimeout` wrapper on HTML5 `play` call to fix issues on mobile browsers ([#694](https://github.com/goldfire/howler.js/pull/694)).
- `FIXED` Remove rare possibility of duplicate sound ID's by using global counter ([#709](https://github.com/goldfire/howler.js/issues/709)).
- `FIXED` Support fades with 2+ decimal places ([#696](https://github.com/goldfire/howler.js/issues/696)).
- `FIXED` Error in Firefox caused by invalid silent WAV on `unload` ([#678](https://github.com/goldfire/howler.js/issues/678)).
- `FIXED` Check for and warn about missing file extension ([#680](https://github.com/goldfire/howler.js/issues/680)).
- `FIXED` Regression in Firefox relating to spatial audio ([#664](https://github.com/goldfire/howler.js/issues/664)).


public/js/main.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var GameObject = (function () {
    function GameObject(el, x, y) {
        this.x = x;
        this.y = y;
        this.div = document.createElement(el);
        document.body.appendChild(this.div);
    }
    GameObject.prototype.update = function () {
    };
    return GameObject;
}());
var Background = (function (_super) {
    __extends(Background, _super);
    function Background(el, x, y) {
        var _this = _super.call(this, el, x, y) || this;
        _this.update();
        return _this;
    }
    Background.prototype.update = function () {
        this.div.style.transform = "translate(" + this.x + "px, " + this.y + "px";
    };
    return Background;
}(GameObject));
var Cat = (function (_super) {
    __extends(Cat, _super);
    function Cat(el, x, y) {
        var _this = _super.call(this, el, x, y) || this;
        _this.div.addEventListener("click", function () { return _this.deleteCat(); });
        _this.update();
        return _this;
    }
    Cat.prototype.deleteCat = function () {
        this.div.remove();
    };
    Cat.prototype.update = function () {
        this.div.style.transform = "translate(" + this.x + "px, " + this.y + "px";
    };
    return Cat;
}(GameObject));
var Counter = (function (_super) {
    __extends(Counter, _super);
    function Counter(el, x, y) {
        var _this = _super.call(this, el, x, y) || this;
        _this.num = 0;
        _this.update();
        return _this;
    }
    Counter.prototype.update = function () {
        this.div.style.transform = "translate(" + this.x + "px, " + this.y + "px";
    };
    return Counter;
}(GameObject));
var Game = (function () {
    function Game() {
        this.screen = new StartMenu(this);
        this.gameLoop();
    }
    Game.prototype.gameLoop = function () {
        var _this = this;
        this.screen.update();
        requestAnimationFrame(function () { return _this.gameLoop(); });
    };
    Game.prototype.emptyScreen = function () {
        document.body.innerHTML = "";
    };
    Game.prototype.showPlayScreen = function (screen) {
        this.screen = screen;
        this.screen.update();
    };
    Game.prototype.showGameOver = function (screen) {
        this.screen = screen;
        this.screen.update();
    };
    Game.prototype.showStartMenu = function (screen) {
        this.screen = screen;
        this.screen.update();
    };
    return Game;
}());
window.addEventListener("load", function () { return new Game(); });
var GameOver = (function () {
    function GameOver(g) {
        var _this = this;
        this.game = g;
        this.background = new Background("gameOverBack", (window.innerWidth / 2) - 657, -8);
        this.element = document.createElement("START");
        document.body.appendChild(this.element);
        this.element.addEventListener("click", function () { return _this.clicked(); });
        this.element.innerHTML = "Game Over<br>" + "You made to many cats!<br>" + "The shelter picked them all up :'(";
    }
    GameOver.prototype.update = function () {
    };
    GameOver.prototype.clicked = function () {
        this.game.emptyScreen();
        this.game.showStartMenu(new StartMenu(this.game));
    };
    return GameOver;
}());
var PlayScreen = (function () {
    function PlayScreen(g) {
        var _this = this;
        this.tooMuchCats = 25;
        this.screenW = window.innerWidth - 320;
        this.screenH = window.innerHeight - 296;
        this.randomNumW = Math.random();
        this.randomNumH = Math.random();
        this.xpos = (this.screenW * this.randomNumW);
        this.ypos = (this.screenH * this.randomNumH);
        this.game = g;
        this.background = new Background("background", -8, -8);
        this.witch = new Witch("witch", 500, 337);
        this.witch.div.addEventListener("click", function () { return _this.counter(); });
        this.count = new Counter("counter", this.screenW / 2, 100);
        this.cat = new Cat("empty", 0, 0);
    }
    PlayScreen.prototype.counter = function () {
        var _this = this;
        this.randomNumW = Math.random();
        this.randomNumH = Math.random();
        this.xpos = (this.screenW * this.randomNumW);
        this.ypos = (this.screenH * this.randomNumH);
        this.cat = new Cat("cat", this.xpos, this.ypos);
        this.cat.div.addEventListener("click", function () { return _this.deleteCat(); });
        this.count.num++;
        console.log("1up");
        this.count.div.innerHTML = ("You have " + this.count.num + " cats");
    };
    PlayScreen.prototype.deleteCat = function () {
        this.count.num--;
        this.count.div.innerHTML = ("You have " + this.count.num + " cats");
    };
    PlayScreen.prototype.clickedSoMuch = function () {
        this.game.emptyScreen();
        this.game.showGameOver(new GameOver(this.game));
    };
    PlayScreen.prototype.update = function () {
        if (this.count.num > this.tooMuchCats) {
            this.clickedSoMuch();
        }
        ;
    };
    return PlayScreen;
}());
var StartMenu = (function () {
    function StartMenu(g) {
        var _this = this;
        this.game = g;
        this.background = new Background("playscreenBack", (window.innerWidth / 2) - 657, -8);
        this.element = document.createElement("start");
        document.body.appendChild(this.element);
        this.element.addEventListener("click", function () { return _this.clicked(); });
        this.element.innerHTML = "Start creating your furry family!";
    }
    StartMenu.prototype.update = function () {
    };
    StartMenu.prototype.clicked = function () {
        this.game.emptyScreen();
        this.game.showPlayScreen(new PlayScreen(this.game));
        console.log("Playscreen started");
    };
    return StartMenu;
}());
var Witch = (function (_super) {
    __extends(Witch, _super);
    function Witch(el, x, y) {
        var _this = _super.call(this, el, x, y) || this;
        _this.update();
        return _this;
    }
    Witch.prototype.update = function () {
        this.div.style.transform = "translate(" + this.x + "px, " + this.y + "px";
    };
    return Witch;
}(GameObject));
//# sourceMappingURL=main.js.map